File: Carros Objective-C/Carros Objective-C/CarroDBCoreData.m
//
//  CarroDBCoreData.m
//  Carros Objective-C
//
//  Created by Antonio Lopes on 8/30/16.
//  Copyright © 2016 Antonio Lopes. All rights reserved.
//


#import "CarroDBCoreData.h"
#import <CoreData/CoreData.h>
#import <UIKit/UIKit.h>
#import "AppDelegate.h"

@implementation CarroDBCoreData

+ (Carro *)newInstance {
    NSManagedObjectContext * context = [CarroDBCoreData getContext];
    Carro * carro = (Carro *)[NSEntityDescription insertNewObjectForEntityForName:@"Carro" inManagedObjectContext:context];
    return carro;
}

+ (NSManagedObjectContext *)getContext {
    AppDelegate * appDelegate = [[UIApplication sharedApplication] delegate];
    return [appDelegate managedObjectContext];
}

- (NSArray<Carro *> *)getCarrosByType:(NSString *)tipo {
    NSManagedObjectContext *context = [CarroDBCoreData getContext];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"Carro" inManagedObjectContext:context];
    
    NSFetchRequest *request = [[NSFetchRequest alloc] init];
    [request setEntity:entity];
    
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"tipo MATCHES %@", tipo];
    [request setPredicate:predicate];

    NSSortDescriptor * sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"timestamp" ascending:true];
    NSArray<NSSortDescriptor *> *sortDescriptors = [NSArray arrayWithObjects:sortDescriptor, nil];
    [request setSortDescriptors:sortDescriptors];
    
    NSError *erro;
    NSArray *array = [context executeFetchRequest:request error:&erro];
    
    if (!array) {
        NSLog(@"Erro ao consultar Carros! %@", [erro localizedDescription]);
        return [NSArray<Carro *> alloc];
    }
    
    return (NSArray<Carro *> *)array;
}

- (void)save:(Carro *)carro {
    if (![carro timestamp]) {
        [carro setTimestamp:[[NSDate alloc] init]];
    }
    
    NSManagedObjectContext *context = [CarroDBCoreData getContext];
    
    NSError *erro = nil;
    [context save:&erro];
    if (!erro) {
        NSLog(@"Carro incluído com sucesso! %@", [carro nome]);
    } else {
        NSLog(@"Erro na inclusão do carro %@! %@", [carro nome], [erro localizedDescription]);
    }
}

- (void)delete:(Carro *)carro {
    NSManagedObjectContext *context = [CarroDBCoreData getContext];
    [context deleteObject:carro];
    
    NSError *erro = nil;
    [context save:&erro];
    if (!erro) {
        NSLog(@"Carro %@ excluído com sucesso!", [carro nome]);
    } else {
        NSLog(@"Erro ao excluir carro %@! %@", [carro nome], [erro localizedDescription]);
    }
}

- (void)deleteCarrosByType:(NSString *)tipo {
    NSArray<Carro *> *carros = [self getCarrosByType:tipo];
    
    while (carros.count > 0) {
        [self delete:[carros objectAtIndex:0]];
    }
}

- (void)close {
    
}

@end

File: Carros Swift/Carros Swift/CarroDBCoreData.swift
//
//  CarroDBCoreData.swift
//  Carros Swift
//
//  Created by Antonio Lopes on 8/30/16.
//  Copyright © 2016 Antonio Lopes. All rights reserved.
//

import Foundation
import CoreData

class CarroDBCoreData {
    
    class func newInstance () -> Carro {
        let context = getContext()
        let c = NSEntityDescription.insertNewObject(forEntityName: "Carro", into: context) as! Carro
        return c
    }
    
    class func getContext() -> NSManagedObjectContext {
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        let context = appDelegate.managedObjectContext
        return context
    }
    
    func getCarrosByType(_ tipo: String) -> Array<Carro> {
        let context = CarroDBCoreData.getContext()
        let entity = NSEntityDescription.entity(forEntityName: "Carro", in: context)
        
        let request = NSFetchRequest<Carro>()
        request.entity = entity
        
        let query = NSPredicate(format: "tipo = \"tipo\"" , argumentArray: nil)
        request.predicate = query
        
        let sortDescriptor = NSSortDescriptor(key: "timestamp", ascending: true)
        let sortDescriptors = [sortDescriptor]
        request.sortDescriptors = sortDescriptors
        
        do {
            let array = try context.fetch(request)
            return array
        } catch {
            print("Erro ao consultar Carros!")
            return [] as Array<Carro>
        }
    }
    
    func save(_ carro: Carro) {
        let context = CarroDBCoreData.getContext()
        
        if (carro.timestamp == nil) {
            carro.timestamp = Date()
        }
        
        do {
            try context.save()
            print("Carro salvo com sucesso!")
        } catch let erro as NSError {
            print("Erro ao salvar: " + erro.localizedDescription)
        }
    }
    
    func delete(_ carro: Carro) {
        let context = CarroDBCoreData.getContext()
        
        context.delete(carro)
        
        do {
            try context.save()
            print("Carro deletado com sucesso!")
        } catch let erro as NSError {
            print("Erro ao deletar: " + erro.localizedDescription)
        }
    }
    
    func deleteCarrosByTipo(_ tipo: String) {
        let carros = getCarrosByType(tipo)
        
        for carro in carros {
            self.delete(carro)
        }
    }
    
    func close() {
        
    }
}
